File: backend/app/schemas/question_editor.py
from __future__ import annotations

import json
import math
import uuid
from datetime import datetime
from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
from typing_extensions import TypeAliasType

from app.core.enums import (
    AnswerPolicy,
    ContentBlockType,
    QuestionDifficulty,
    QuestionRevisionStatus,
)
from app.schemas.question_answer import AcceptedAnswerRead, AnswerOptionRead
from app.schemas.question_solution import SolutionRead
from app.schemas.structured_text import StructuredTextDocument


class StrictEditorSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


JsonValue = TypeAliasType(
    "JsonValue",
    Union[
        str,
        int,
        float,
        bool,
        None,
        list["JsonValue"],
        dict[str, "JsonValue"],
    ],
)

_GEOMETRY_MAX_JSON_BYTES = 1_048_576
_GEOMETRY_MAX_JSON_DEPTH = 32


def _validate_geometry_source_data(value: object) -> object:
    """Require JSON values; the outer object has container depth one."""

    if not isinstance(value, dict):
        raise ValueError("Geometry source_data must be a JSON object.")
    stack: list[tuple[object, int]] = [(value, 1)]
    while stack:
        current, depth = stack.pop()
        if isinstance(current, dict):
            if depth > _GEOMETRY_MAX_JSON_DEPTH:
                raise ValueError("Geometry source_data exceeds maximum depth.")
            for key, nested in current.items():
                if not isinstance(key, str):
                    raise ValueError("Geometry JSON object keys must be strings.")
                stack.append((nested, depth + 1))
        elif isinstance(current, list):
            if depth > _GEOMETRY_MAX_JSON_DEPTH:
                raise ValueError("Geometry source_data exceeds maximum depth.")
            stack.extend((nested, depth + 1) for nested in current)
        elif isinstance(current, float):
            if not math.isfinite(current):
                raise ValueError("Geometry source_data numbers must be finite.")
        elif current is None or isinstance(current, (str, int, bool)):
            continue
        else:
            raise ValueError("Geometry source_data must contain only JSON values.")

    try:
        encoded = json.dumps(
            value,
            ensure_ascii=False,
            separators=(",", ":"),
            allow_nan=False,
        ).encode("utf-8")
    except (TypeError, ValueError, OverflowError) as exc:
        raise ValueError("Geometry source_data is not valid JSON.") from exc
    if len(encoded) > _GEOMETRY_MAX_JSON_BYTES:
        raise ValueError("Geometry source_data exceeds maximum encoded size.")
    return value


def _require_unique_ids(values: list[uuid.UUID], field_name: str) -> list[uuid.UUID]:
    if len(values) != len(set(values)):
        raise ValueError(f"{field_name} must contain unique IDs.")
    return values


def _require_aware_datetime(value: datetime) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("Concurrency timestamp must include a timezone.")
    return value


class QuestionDraftCreate(StrictEditorSchema):
    question_type_id: uuid.UUID
    primary_topic_id: uuid.UUID | None = None
    related_topic_ids: list[uuid.UUID] = Field(default_factory=list)
    purpose_ids: list[uuid.UUID] = Field(default_factory=list)

    @field_validator("related_topic_ids")
    @classmethod
    def validate_related_topic_ids(
        cls, values: list[uuid.UUID],
    ) -> list[uuid.UUID]:
        return _require_unique_ids(values, "related_topic_ids")

    @field_validator("purpose_ids")
    @classmethod
    def validate_purpose_ids(
        cls, values: list[uuid.UUID],
    ) -> list[uuid.UUID]:
        return _require_unique_ids(values, "purpose_ids")

    @model_validator(mode="after")
    def reject_primary_topic_as_related(self) -> "QuestionDraftCreate":
        if (
            self.primary_topic_id is not None
            and self.primary_topic_id in self.related_topic_ids
        ):
            raise ValueError("Primary topic cannot also be a related topic.")
        return self


class QuestionDraftRead(StrictEditorSchema):
    question_family_id: uuid.UUID
    question_form_id: uuid.UUID
    revision_id: uuid.UUID
    revision_number: int = Field(gt=0)
    status: QuestionRevisionStatus
    question_type_id: uuid.UUID
    source_id: uuid.UUID | None
    source_detail: str | None
    source_display_name: str | None
    primary_topic_id: uuid.UUID | None
    related_topic_ids: list[uuid.UUID]
    purpose_ids: list[uuid.UUID]
    difficulty: QuestionDifficulty | None
    updated_at: datetime

    @field_validator("updated_at")
    @classmethod
    def validate_updated_at(cls, value: datetime) -> datetime:
        return _require_aware_datetime(value)


class TextBlockPayloadRead(StrictEditorSchema):
    source_text: str
    document: StructuredTextDocument
    format_version: Literal[1]


class FormulaBlockPayloadRead(StrictEditorSchema):
    source_latex: str
    format_version: Literal[1]


class ImageBlockPayloadRead(StrictEditorSchema):
    media_asset_id: uuid.UUID
    alt_text: str | None


class GeometryBlockPayloadRead(StrictEditorSchema):
    source_data: dict[str, object]
    format_version: Literal[1]


class TextBlockRead(StrictEditorSchema):
    id: uuid.UUID
    block_type: Literal[ContentBlockType.TEXT]
    sort_order: int = Field(ge=0)
    payload: TextBlockPayloadRead


class FormulaBlockRead(StrictEditorSchema):
    id: uuid.UUID
    block_type: Literal[ContentBlockType.FORMULA]
    sort_order: int = Field(ge=0)
    payload: FormulaBlockPayloadRead


class ImageBlockRead(StrictEditorSchema):
    id: uuid.UUID
    block_type: Literal[ContentBlockType.IMAGE]
    sort_order: int = Field(ge=0)
    payload: ImageBlockPayloadRead


class GeometryBlockRead(StrictEditorSchema):
    id: uuid.UUID
    block_type: Literal[ContentBlockType.GEOMETRY]
    sort_order: int = Field(ge=0)
    payload: GeometryBlockPayloadRead


ContentBlockRead = Annotated[
    Union[TextBlockRead, FormulaBlockRead, ImageBlockRead, GeometryBlockRead],
    Field(discriminator="block_type"),
]


class QuestionRevisionEditorRead(QuestionDraftRead):
    blocks: list[ContentBlockRead]
    answer_policy: AnswerPolicy = AnswerPolicy.UNSUPPORTED
    answer_options: list[AnswerOptionRead] = Field(default_factory=list)
    accepted_answers: list[AcceptedAnswerRead] = Field(default_factory=list)
    solution: SolutionRead | None = None


class TextBlockWritePayload(StrictEditorSchema):
    document: StructuredTextDocument
    format_version: Literal[1] = 1


class FormulaBlockWritePayload(StrictEditorSchema):
    source_latex: str
    format_version: Literal[1] = 1


class ImageBlockWritePayload(StrictEditorSchema):
    media_asset_id: uuid.UUID
    alt_text: str | None


class GeometryBlockWritePayload(StrictEditorSchema):
    source_data: dict[str, JsonValue]
    format_version: Literal[1] = 1

    @field_validator("source_data", mode="before")
    @classmethod
    def validate_source_data(cls, value: object) -> object:
        return _validate_geometry_source_data(value)


class TextBlockCreate(StrictEditorSchema):
    block_type: Literal[ContentBlockType.TEXT]
    payload: TextBlockWritePayload
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class FormulaBlockCreate(StrictEditorSchema):
    block_type: Literal[ContentBlockType.FORMULA]
    payload: FormulaBlockWritePayload
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class ImageBlockCreate(StrictEditorSchema):
    block_type: Literal[ContentBlockType.IMAGE]
    payload: ImageBlockWritePayload
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class GeometryBlockCreate(StrictEditorSchema):
    block_type: Literal[ContentBlockType.GEOMETRY]
    payload: GeometryBlockWritePayload
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


ContentBlockCreate = Annotated[
    Union[
        TextBlockCreate,
        FormulaBlockCreate,
        ImageBlockCreate,
        GeometryBlockCreate,
    ],
    Field(discriminator="block_type"),
]


class TextBlockUpdate(StrictEditorSchema):
    document: StructuredTextDocument
    format_version: Literal[1] = 1
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class FormulaBlockUpdate(StrictEditorSchema):
    source_latex: str
    format_version: Literal[1] = 1
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class ImageBlockUpdate(StrictEditorSchema):
    media_asset_id: uuid.UUID
    alt_text: str | None
    expected_revision_updated_at: datetime

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


class GeometryBlockUpdate(StrictEditorSchema):
    source_data: dict[str, JsonValue]
    format_version: Literal[1] = 1
    expected_revision_updated_at: datetime

    @field_validator("source_data", mode="before")
    @classmethod
    def validate_source_data(cls, value: object) -> object:
        return _validate_geometry_source_data(value)

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)


ContentBlockUpdate = Union[
    TextBlockUpdate,
    FormulaBlockUpdate,
    ImageBlockUpdate,
    GeometryBlockUpdate,
]


class BlockOrderRequest(StrictEditorSchema):
    block_ids: list[uuid.UUID]
    expected_revision_updated_at: datetime

    @field_validator("block_ids")
    @classmethod
    def validate_block_ids(
        cls, values: list[uuid.UUID],
    ) -> list[uuid.UUID]:
        return _require_unique_ids(values, "block_ids")

    @field_validator("expected_revision_updated_at")
    @classmethod
    def validate_expected_revision_updated_at(
        cls, value: datetime,
    ) -> datetime:
        return _require_aware_datetime(value)

File: frontend/src/api/questionEditor.ts
import { requestJson } from './client'

export type UUID = string
export type IsoDateTime = string

export type QuestionRevisionStatus =
  | 'draft'
  | 'proposed'
  | 'approved'
  | 'rejected'

export type QuestionDifficulty = 'easy' | 'medium' | 'hard'
export type AnswerPolicy = 'option_single' | 'option_multiple'
  | 'accepted_answer' | 'none' | 'unsupported'

export type BoldMark = {
  type: 'bold'
}

export type ItalicMark = {
  type: 'italic'
}

export type UnderlineMark = {
  type: 'underline'
}

export type FontFamilyMark = {
  type: 'font_family'
  value: 'default' | 'serif' | 'sans' | 'math-compatible'
}

export type FontSizeMark = {
  type: 'font_size'
  value: 'small' | 'normal' | 'large' | 'x-large'
}

export type TextMark =
  | BoldMark
  | ItalicMark
  | UnderlineMark
  | FontFamilyMark
  | FontSizeMark

export type TextNode = {
  type: 'text'
  text: string
  marks: TextMark[]
}

export type InlineMathNode = {
  type: 'inline_math'
  latex: string
}

export type HardBreakNode = {
  type: 'hard_break'
}

export type InlineNode = TextNode | InlineMathNode | HardBreakNode

export type ParagraphAttrs = {
  alignment: 'start' | 'center' | 'end' | 'justify'
}

export type ParagraphNode = {
  type: 'paragraph'
  attrs: ParagraphAttrs | null
  content: InlineNode[]
}

export type ListItemNode = {
  type: 'list_item'
  content: ParagraphNode[]
}

export type BulletListNode = {
  type: 'bullet_list'
  content: ListItemNode[]
}

export type OrderedListNode = {
  type: 'ordered_list'
  content: ListItemNode[]
}

export type StructuredTextBlockNode =
  | ParagraphNode
  | BulletListNode
  | OrderedListNode

export type StructuredTextDocument = {
  type: 'document'
  content: StructuredTextBlockNode[]
}

export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue }

export type JsonObject = { [key: string]: JsonValue }

export type QuestionDraftCreate = {
  question_type_id: UUID
  primary_topic_id?: UUID | null
  related_topic_ids?: UUID[]
  purpose_ids?: UUID[]
}

export type QuestionDraftRead = {
  question_family_id: UUID
  question_form_id: UUID
  revision_id: UUID
  revision_number: number
  status: QuestionRevisionStatus
  question_type_id: UUID
  source_id: UUID | null
  source_detail: string | null
  source_display_name: string | null
  primary_topic_id: UUID | null
  related_topic_ids: UUID[]
  purpose_ids: UUID[]
  difficulty: QuestionDifficulty | null
  updated_at: IsoDateTime
}

export type TextBlockPayloadRead = {
  source_text: string
  document: StructuredTextDocument
  format_version: 1
}

export type TextBlockWritePayload = {
  document: StructuredTextDocument
  format_version?: 1
}

export type TextBlockCreate = {
  block_type: 'text'
  payload: TextBlockWritePayload
  expected_revision_updated_at: IsoDateTime
}

export type TextBlockUpdate = {
  document: StructuredTextDocument
  format_version?: 1
  expected_revision_updated_at: IsoDateTime
}

export type FormulaBlockPayloadRead = {
  source_latex: string
  format_version: 1
}

export type FormulaBlockWritePayload = {
  source_latex: string
  format_version?: 1
}

export type FormulaBlockCreate = {
  block_type: 'formula'
  payload: FormulaBlockWritePayload
  expected_revision_updated_at: IsoDateTime
}

export type FormulaBlockUpdate = {
  source_latex: string
  format_version?: 1
  expected_revision_updated_at: IsoDateTime
}

export type ImageBlockPayloadRead = {
  media_asset_id: UUID
  alt_text: string | null
}

export type ImageBlockCreate = {
  block_type: 'image'
  payload: ImageBlockPayloadRead
  expected_revision_updated_at: IsoDateTime
}

export type ImageBlockUpdate = ImageBlockPayloadRead & {
  expected_revision_updated_at: IsoDateTime
}

export type GeometryBlockPayloadRead = {
  source_data: JsonObject
  format_version: 1
}

export type GeometryBlockWritePayload = {
  source_data: JsonObject
  format_version?: 1
}

export type GeometryBlockCreate = {
  block_type: 'geometry'
  payload: GeometryBlockWritePayload
  expected_revision_updated_at: IsoDateTime
}

export type GeometryBlockUpdate = {
  source_data: JsonObject
  format_version?: 1
  expected_revision_updated_at: IsoDateTime
}

export type BlockDeleteRequest = {
  expected_revision_updated_at: IsoDateTime
}

export type BlockOrderRequest = {
  block_ids: UUID[]
  expected_revision_updated_at: IsoDateTime
}

export type TextBlockRead = {
  id: UUID
  block_type: 'text'
  sort_order: number
  payload: TextBlockPayloadRead
}

export type FormulaBlockRead = {
  id: UUID
  block_type: 'formula'
  sort_order: number
  payload: FormulaBlockPayloadRead
}

export type ImageBlockRead = {
  id: UUID
  block_type: 'image'
  sort_order: number
  payload: ImageBlockPayloadRead
}

export type GeometryBlockRead = {
  id: UUID
  block_type: 'geometry'
  sort_order: number
  payload: GeometryBlockPayloadRead
}

export type ContentBlockRead =
  | TextBlockRead
  | FormulaBlockRead
  | ImageBlockRead
  | GeometryBlockRead

export type SolutionTextBlockRead = {
  id: UUID
  block_type: 'text'
  sort_order: number
  source_text: string
  document: StructuredTextDocument
  format_version: 1
}

export type SolutionFormulaBlockRead = {
  id: UUID
  block_type: 'formula'
  sort_order: number
  source_latex: string
  format_version: 1
}

export type SolutionBlockRead = SolutionTextBlockRead | SolutionFormulaBlockRead
export type SolutionRead = { id: UUID; blocks: SolutionBlockRead[] }
export type SolutionMutationRequest = { expected_revision_updated_at: IsoDateTime }
export type SolutionTextBlockCreate = SolutionMutationRequest & {
  block_type: 'text'
  payload: TextBlockWritePayload
}
export type SolutionTextBlockUpdate = SolutionMutationRequest & { payload: TextBlockWritePayload }
export type SolutionFormulaBlockCreate = SolutionMutationRequest & {
  block_type: 'formula'
  payload: FormulaBlockWritePayload
}
export type SolutionFormulaBlockUpdate = SolutionMutationRequest & { payload: FormulaBlockWritePayload }
export type SolutionBlockOrderRequest = SolutionMutationRequest & { block_ids: UUID[] }

export type QuestionRevisionEditorRead = QuestionDraftRead & {
  blocks: ContentBlockRead[]
  answer_policy: AnswerPolicy
  answer_options: AnswerOptionRead[]
  accepted_answers: AcceptedAnswerRead[]
  solution: SolutionRead | null
}

export type AnswerOptionRead = {
  id: UUID
  label: string | null
  order_index: number
  source_text: string
  document: StructuredTextDocument
  format_version: 1
  is_correct: boolean
}

export type AcceptedAnswerRead = {
  id: UUID
  order_index: number
  source_text: string
  document: StructuredTextDocument
  format_version: 1
}

export type AnswerContentRequest = {
  document: StructuredTextDocument
  format_version?: 1
  expected_revision_updated_at: IsoDateTime
}

export type AnswerOptionRequest = AnswerContentRequest & { label?: string | null }
export type AnswerOrderRequest = { answer_ids: UUID[]; expected_revision_updated_at: IsoDateTime }
export type SetCorrectOptionsRequest = { option_ids: UUID[]; expected_revision_updated_at: IsoDateTime }

export function createQuestionDraft(
  accessToken: string,
  request: QuestionDraftCreate,
): Promise<QuestionDraftRead> {
  return requestJson<QuestionDraftRead>('/api/v1/question-editor/drafts', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  })
}

export function getQuestionRevisionForEditor(
  accessToken: string,
  revisionId: UUID,
): Promise<QuestionRevisionEditorRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<QuestionRevisionEditorRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}`,
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    },
  )
}

export function createTextBlock(
  accessToken: string,
  revisionId: UUID,
  request: TextBlockCreate,
): Promise<TextBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<TextBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/text`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function updateTextBlock(
  accessToken: string,
  revisionId: UUID,
  blockId: UUID,
  request: TextBlockUpdate,
): Promise<TextBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)
  const encodedBlockId = encodeURIComponent(blockId)

  return requestJson<TextBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/`
      + `${encodedBlockId}/text`,
    {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function createFormulaBlock(
  accessToken: string,
  revisionId: UUID,
  request: FormulaBlockCreate,
): Promise<FormulaBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<FormulaBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/formula`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function updateFormulaBlock(
  accessToken: string,
  revisionId: UUID,
  blockId: UUID,
  request: FormulaBlockUpdate,
): Promise<FormulaBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)
  const encodedBlockId = encodeURIComponent(blockId)

  return requestJson<FormulaBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/`
      + `${encodedBlockId}/formula`,
    {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function createImageBlock(
  accessToken: string,
  revisionId: UUID,
  request: ImageBlockCreate,
): Promise<ImageBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<ImageBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/image`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function updateImageBlock(
  accessToken: string,
  revisionId: UUID,
  blockId: UUID,
  request: ImageBlockUpdate,
): Promise<ImageBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)
  const encodedBlockId = encodeURIComponent(blockId)

  return requestJson<ImageBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/`
      + `${encodedBlockId}/image`,
    {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function createGeometryBlock(
  accessToken: string,
  revisionId: UUID,
  request: GeometryBlockCreate,
): Promise<GeometryBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<GeometryBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/geometry`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function updateGeometryBlock(
  accessToken: string,
  revisionId: UUID,
  blockId: UUID,
  request: GeometryBlockUpdate,
): Promise<GeometryBlockRead> {
  const encodedRevisionId = encodeURIComponent(revisionId)
  const encodedBlockId = encodeURIComponent(blockId)

  return requestJson<GeometryBlockRead>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/`
      + `${encodedBlockId}/geometry`,
    {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

export function deleteBlock(
  accessToken: string,
  revisionId: UUID,
  blockId: UUID,
  request: BlockDeleteRequest,
): Promise<void> {
  const encodedRevisionId = encodeURIComponent(revisionId)
  const encodedBlockId = encodeURIComponent(blockId)
  const encodedTimestamp = encodeURIComponent(
    request.expected_revision_updated_at,
  )

  return requestJson<void>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/`
      + `${encodedBlockId}?expected_revision_updated_at=${encodedTimestamp}`,
    {
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    },
  )
}

export function reorderBlocks(
  accessToken: string,
  revisionId: UUID,
  request: BlockOrderRequest,
): Promise<void> {
  const encodedRevisionId = encodeURIComponent(revisionId)

  return requestJson<void>(
    `/api/v1/question-editor/revisions/${encodedRevisionId}/blocks/order`,
    {
      method: 'PUT',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )
}

function answerPath(revisionId: UUID, collection: 'answer-options' | 'accepted-answers'): string {
  return `/api/v1/question-editor/revisions/${encodeURIComponent(revisionId)}/${collection}`
}

export function createOption(accessToken: string, revisionId: UUID, request: AnswerOptionRequest): Promise<AnswerOptionRead> {
  return requestJson<AnswerOptionRead>(answerPath(revisionId, 'answer-options'), { method: 'POST', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function updateOption(accessToken: string, revisionId: UUID, optionId: UUID, request: AnswerOptionRequest): Promise<AnswerOptionRead> {
  return requestJson<AnswerOptionRead>(`${answerPath(revisionId, 'answer-options')}/${encodeURIComponent(optionId)}`, { method: 'PATCH', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function deleteOption(accessToken: string, revisionId: UUID, optionId: UUID, expectedAt: IsoDateTime): Promise<void> {
  return requestJson<void>(`${answerPath(revisionId, 'answer-options')}/${encodeURIComponent(optionId)}?expected_revision_updated_at=${encodeURIComponent(expectedAt)}`, { method: 'DELETE', headers: { Authorization: `Bearer ${accessToken}` } })
}

export function reorderOptions(accessToken: string, revisionId: UUID, request: AnswerOrderRequest): Promise<AnswerOptionRead[]> {
  return requestJson<AnswerOptionRead[]>(`${answerPath(revisionId, 'answer-options')}/actions/order`, { method: 'PUT', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function setCorrectOptions(accessToken: string, revisionId: UUID, request: SetCorrectOptionsRequest): Promise<AnswerOptionRead[]> {
  return requestJson<AnswerOptionRead[]>(`${answerPath(revisionId, 'answer-options')}/actions/correct`, { method: 'PUT', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function createAcceptedAnswer(accessToken: string, revisionId: UUID, request: AnswerContentRequest): Promise<AcceptedAnswerRead> {
  return requestJson<AcceptedAnswerRead>(answerPath(revisionId, 'accepted-answers'), { method: 'POST', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function updateAcceptedAnswer(accessToken: string, revisionId: UUID, answerId: UUID, request: AnswerContentRequest): Promise<AcceptedAnswerRead> {
  return requestJson<AcceptedAnswerRead>(`${answerPath(revisionId, 'accepted-answers')}/${encodeURIComponent(answerId)}`, { method: 'PATCH', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

export function deleteAcceptedAnswer(accessToken: string, revisionId: UUID, answerId: UUID, expectedAt: IsoDateTime): Promise<void> {
  return requestJson<void>(`${answerPath(revisionId, 'accepted-answers')}/${encodeURIComponent(answerId)}?expected_revision_updated_at=${encodeURIComponent(expectedAt)}`, { method: 'DELETE', headers: { Authorization: `Bearer ${accessToken}` } })
}

export function reorderAcceptedAnswers(accessToken: string, revisionId: UUID, request: AnswerOrderRequest): Promise<AcceptedAnswerRead[]> {
  return requestJson<AcceptedAnswerRead[]>(`${answerPath(revisionId, 'accepted-answers')}/actions/order`, { method: 'PUT', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request) })
}

function solutionPath(revisionId: UUID): string {
  return `/api/v1/question-editor/revisions/${encodeURIComponent(revisionId)}/solution`
}

export function getSolution(accessToken: string, revisionId: UUID): Promise<SolutionRead | null> {
  return requestJson<SolutionRead | null>(solutionPath(revisionId), {
    method: 'GET', headers: { Authorization: `Bearer ${accessToken}` },
  })
}

export function createSolution(accessToken: string, revisionId: UUID, request: SolutionMutationRequest): Promise<SolutionRead> {
  return requestJson<SolutionRead>(solutionPath(revisionId), {
    method: 'POST', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function deleteSolution(accessToken: string, revisionId: UUID, request: SolutionMutationRequest): Promise<void> {
  return requestJson<void>(solutionPath(revisionId), {
    method: 'DELETE', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function createSolutionTextBlock(accessToken: string, revisionId: UUID, request: SolutionTextBlockCreate): Promise<SolutionTextBlockRead> {
  return requestJson<SolutionTextBlockRead>(`${solutionPath(revisionId)}/blocks/text`, {
    method: 'POST', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function updateSolutionTextBlock(accessToken: string, revisionId: UUID, blockId: UUID, request: SolutionTextBlockUpdate): Promise<SolutionTextBlockRead> {
  return requestJson<SolutionTextBlockRead>(`${solutionPath(revisionId)}/blocks/${encodeURIComponent(blockId)}/text`, {
    method: 'PATCH', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function createSolutionFormulaBlock(accessToken: string, revisionId: UUID, request: SolutionFormulaBlockCreate): Promise<SolutionFormulaBlockRead> {
  return requestJson<SolutionFormulaBlockRead>(`${solutionPath(revisionId)}/blocks/formula`, {
    method: 'POST', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function updateSolutionFormulaBlock(accessToken: string, revisionId: UUID, blockId: UUID, request: SolutionFormulaBlockUpdate): Promise<SolutionFormulaBlockRead> {
  return requestJson<SolutionFormulaBlockRead>(`${solutionPath(revisionId)}/blocks/${encodeURIComponent(blockId)}/formula`, {
    method: 'PATCH', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

export function deleteSolutionBlock(accessToken: string, revisionId: UUID, blockId: UUID, expectedAt: IsoDateTime): Promise<void> {
  return requestJson<void>(`${solutionPath(revisionId)}/blocks/${encodeURIComponent(blockId)}?expected_revision_updated_at=${encodeURIComponent(expectedAt)}`, {
    method: 'DELETE', headers: { Authorization: `Bearer ${accessToken}` },
  })
}

export function reorderSolutionBlocks(accessToken: string, revisionId: UUID, request: SolutionBlockOrderRequest): Promise<SolutionBlockRead[]> {
  return requestJson<SolutionBlockRead[]>(`${solutionPath(revisionId)}/blocks/actions/order`, {
    method: 'PUT', headers: { Authorization: `Bearer ${accessToken}`, 'Content-Type': 'application/json' }, body: JSON.stringify(request),
  })
}

